test/kernels.jl

```diff
@@ -85,11 +85,17 @@ function record(f, input...)
     return tape, recorded_output, recorded_input
 end
 
-function autograd(f, input...)
+function autograd(f, input...; cache::Bool = false)
     tape, recorded_output, recorded_input = record(f, input...)
     forward!(tape)
     seed!(recorded_output)
     backward!(tape)
+    if cache
```

Comment:
Should this note be:
```
if cache
    initderiv!(tape)
end
forward!(tape)
seed!(recorded_output)
backward!(tape)
```

So that `cache==true` doesn't end up running the tape twice?

Reply:
No, the code is correct AFAICT (all `initderiv!` does is zero out derivative buffers in the tape).

All this PR is doing is adding a check to see if reusable buffers were allocated during previous tape passes. To test the code paths used when this check returns true, we need to run the tape at least once before executing the "actual" passes.